event_model/__init__.py

```diff
@@ -1641,109 +1818,176 @@ class ComposeRunBundle:
     run bundle (start, compose_descriptor, compose_resource, stop). Further extensions are optional and
     require keyword referencing (i.e. compose_stream_resource).
     """
-    start_doc: dict
-    compose_descriptor: callable
-    compose_resource: callable
-    compose_stop: callable
-    compose_stream_resource: Optional[callable] = None
-
-    def __iter__(self):
-        return iter((self.start_doc, self.compose_descriptor, self.compose_resource, self.compose_stop))
-
 
-ComposeDescriptorBundle = namedtuple('ComposeDescriptorBundle',
-                                     'descriptor_doc compose_event compose_event_page')
-ComposeResourceBundle = namedtuple('ComposeResourceBundle',
-                                   'resource_doc compose_datum compose_datum_page')
-ComposeStreamResourceBundle = namedtuple('ComposeStreamResourceBundle',
-                                         'stream_resource_doc compose_stream_data')
+    start_doc: dict
+    compose_descriptor: Callable
+    compose_resource: Callable
+    compose_stop: Callable
+    compose_stream_resource: Optional[Callable] = None
+
+    def __iter__(self) -> Iterator:
+        return iter(
+            (
+                self.start_doc,
+                self.compose_descriptor,
+                self.compose_resource,
+                self.compose_stop,
+            )
+        )
 
 
-def compose_datum(*, resource, counter, datum_kwargs, validate=True):
-    resource_uid = resource['uid']
-    doc = {'resource': resource_uid,
-           'datum_kwargs': datum_kwargs,
-           'datum_id': '{}/{}'.format(resource_uid, next(counter))}
+ComposeDescriptorBundle = namedtuple(
+    "ComposeDescriptorBundle", "descriptor_doc compose_event compose_event_page"
+)
+ComposeResourceBundle = namedtuple(
+    "ComposeResourceBundle", "resource_doc compose_datum compose_datum_page"
+)
+ComposeStreamResourceBundle = namedtuple(
+    "ComposeStreamResourceBundle", "stream_resource_doc compose_stream_data"
+)
+
+
+def compose_datum(
+    *,
+    resource: dict,
+    counter: Iterator,
+    datum_kwargs: Iterable,
+    validate: bool = True,
+) -> dict:
```

Comment:
Please can we tie down the return values from `compose_*` a bit more tightly, I'd like to see a `documents.py` that contains `TypedDict`s that are returned from these functions.

A bit like
https://github.com/bluesky/bluesky/blob/d2b6bddd4415cdb2946fe8d552bd704bc9932c11/bluesky/protocols.py#L13-L30